packages/runtime/src/runtime/hooks/useEnsembleAction.tsx

```diff
@@ -299,6 +315,179 @@ export const useInvokeAPI: EnsembleActionHook<InvokeAPIAction> = (action) => {
+        };
+
+        ws.onmessage = (e: MessageEvent): void => {
+          onMessageReceiveAction?.callback({ data: e.data as unknown });
+        };
+
+        ws.onclose = (): void => {
+          onSocketDisconnectAction?.callback();
+        };
+
+        setData(socket.name, ws);
+      } catch (e) {
+        logError(e);
+      } finally {
+        setIsLoading(false);
+        setIsComplete(true);
+      }
+    };
+
+    fireRequest();
+  }, [
+    socket,
+    isComplete,
+    isLoading,
+    setData,
+    onSocketConnectAction,
+    onMessageReceiveAction,
+    onSocketDisconnectAction,
+  ]);
+
+  return connectSocket;
+};
+
+export const useMessageSocket: EnsembleActionHook<SendSocketMessageAction> = (
+  action,
+) => {
+  const { sockets, data } = useScreenData();
+  const [isComplete, setIsComplete] = useState<boolean>();
+  const [isLoading, setIsLoading] = useState<boolean>();
+  const [context, setContext] = useState<{ [key: string]: unknown }>();
+  const evaluatedInputs = useEvaluate(action?.message, { context });
+
+  const socket = useMemo(
+    () => sockets?.find((model) => model.name === action?.name),
+    [action, sockets],
+  );
+
+  const sendSocketMessage = useMemo(() => {
+    if (!socket) {
+      return;
+    }
+
+    const callback = (args: unknown): void => {
+      setIsComplete(false);
+      setContext(args as { [key: string]: unknown });
+    };
+    return { callback };
+  }, [socket]);
+
+  useEffect(() => {
+    if (!socket || isComplete !== false || isLoading) {
+      return;
+    }
+
+    const fireRequest = (): void => {
+      setIsLoading(true);
+      try {
+        const socketInstance = data[socket.name] as WebSocket;
+        if (socketInstance) {
+          socketInstance.send(JSON.stringify(evaluatedInputs));
+        }
+      } catch (e) {
+        logError(e);
+      } finally {
+        setIsLoading(false);
+        setIsComplete(true);
+      }
+    };
+
+    fireRequest();
+  }, [socket, data, evaluatedInputs, isComplete, isLoading]);
+
+  return sendSocketMessage;
+};
+
+export const useDisconnectSocket: EnsembleActionHook<DisconnectSocketAction> = (
+  action,
+) => {
+  const { sockets, data, setData } = useScreenData();
+  const [isComplete, setIsComplete] = useState<boolean>();
+  const [isLoading, setIsLoading] = useState<boolean>();
+
+  const socket = useMemo(
+    () => sockets?.find((model) => model.name === action?.name),
+    [action, sockets],
+  );
+
+  const disconnectSocket = useMemo(() => {
+    if (!socket) {
+      return;
+    }
+
+    const callback = (): void => {
+      setIsComplete(false);
+    };
+    return { callback };
+  }, [socket]);
+
+  useEffect(() => {
+    if (!socket || isComplete !== false || isLoading) {
+      return;
+    }
+
+    const fireRequest = (): void => {
+      setIsLoading(true);
+      try {
+        const socketInstance = data[socket.name] as WebSocket;
+        if (socketInstance) {
+          socketInstance.close();
```

Comment:
Awaitable?

Reply:
No, this method will close the socket by closing handshake after checking there is no message left in queue. once the queue will be empty then closing handshake will begin and close the socket